src/taskorchestrationexecutor.ts

```diff
@@ -0,0 +1,493 @@
+    private deferredTasks: Record<number | string, () => void>;
+    private sequenceNumber: number;
+    private schemaVersion: UpperSchemaVersion;
+    public willContinueAsNew: boolean;
+    private actions: IAction[];
+    protected openTasks: Record<number | string, TaskBase>;
+    protected openEvents: Record<number | string, TaskBase[]>;
+    private eventToTaskValuePayload: { [key in HistoryEventType]?: [boolean, string] };
+
+    constructor() {
+        // Map of task-completion events types to pairs of
+        // (1) whether that event corresponds to a successful task result, and
+        // (2) the field in the event type that would contain the task's ID.
+        this.eventToTaskValuePayload = {
+            [HistoryEventType.TaskCompleted]: [true, "TaskScheduledId"],
+            [HistoryEventType.TimerFired]: [true, "TimerId"],
+            [HistoryEventType.SubOrchestrationInstanceCompleted]: [true, "TaskScheduledId"],
+            [HistoryEventType.EventRaised]: [true, "Name"],
+            [HistoryEventType.TaskFailed]: [false, "TaskScheduledId"],
+            [HistoryEventType.SubOrchestrationInstanceFailed]: [false, "TaskScheduledId"],
+        };
+        this.initialize();
+    }
+
+    /**
+     * @hidden
+     *
+     * Initialize the task orchestration executor for a brand new orchestrator invocation.
+     * To be called in ContinueAsNew scenarios as well.
+     */
+    private initialize(): void {
+        // The very first task, to kick-start the generator, is just a dummy/no-op task
+        this.currentTask = new InternalOnlyTask();
+        this.currentTask.SetValue(false, undefined);
+
+        this.sequenceNumber = 0;
+        this.willContinueAsNew = false;
+        this.openTasks = {};
+        this.openEvents = {};
+        this.actions = [];
+        this.deferredTasks = {};
+
+        this.output = undefined;
+        this.exception = undefined;
+        this.orchestratorReturned = false;
+    }
+
+    /**
+     * @hidden
+     *
+     * Start an orchestration's execution, replaying based on the currently-available History.
+     *
+     * @param context
+     *  The orchestration context
+     * @param history
+     *  The orchestration history
+     * @param schemaVersion
+     *  The OOProc output schema version expected by the DF extension
+     * @param fn
+     *  The user-code defining the orchestration
+     *
+     * @returns
+     *  Returns void but communicates the resulting orchestrator state via the context object's handler
+     */
+    public async execute(
+        context: IOrchestrationFunctionContext,
+        history: HistoryEvent[],
+        schemaVersion: UpperSchemaVersion,
+        fn: (context: IOrchestrationFunctionContext) => IterableIterator<unknown>
+    ): Promise<void> {
+        this.schemaVersion = schemaVersion;
+        this.context = context.df;
+        this.generator = fn(context) as Generator<TaskBase, any, any>; // TODO: what happens if code is not a generator?
```

Comment:
Regarding this TODO, should we throw an error if it's not a generator?

Reply:
We actually support non-generators in Python. For example, a trivial orchestration that just returns "Hello!" is a valid orchestration.

Reply:
Oh interesting. I guess if we wanted to also support that in Python we could just set `fn` as the orchestrator output value if it's not a generator.

Reply:
I tested this on a newly added unit test, and it seems the current implementation already supports non-generator orchestrations; which is exactly what we want!

I've removed this TODO comment